# Diabetes Exploratory Data Analysis



__Author:__ Desiree Unselt <br>
__Date:__ 24JUL2024 <br>

__Dataset:__ Diabetes 130 dataset representing 130 US hospitals collected from 1999 to 2008, hosted on the UC Irvine Dataset Repository.

__Objectives:__
  - Perform an exploratory data analysis (EDA) on the dataset to understand how it's structured and what it contains
  - Once completed:
    - Ask someone to review your code
    - Share your findings wiht the team
    - Commit your code to a local git repository
  - Think about how you could train a model to predict patient readmission, then try and build a model to predict this property.
    - The dataset has the following classes in the readmission column:
        - 'No' (didn't get readmitted)
        - '>30' (was readmitted after 30 days)
        - '<30' (was readmitted within 30 days)
    - How can you measure the performance of the model? How well is it doing?
    - What's the best model that you can develop?
    - How does it compare against other volunteer models? Can you combine ideas from other people to make your model better?
  - Can you come up with a reasonable real life application of these algorithms being applied to the Diabetes data? Are there any limitations of your approach?

## Exploratory Data Analysis (EDA)

### Import Libraries, Set Working Directory, and Load Data

In [29]:
import importlib

packages = ['sklearn', 'seaborn', 'matplotlib', 'scipy']

for package in packages:
    try:
        importlib.import_module(package)
        print(f"{package} is installed")
    except ImportError:
        print(f"{package} is NOT installed")

sklearn is NOT installed
seaborn is NOT installed
matplotlib is installed
scipy is NOT installed


In [30]:
# Import libraries 
import os
import pandas as pd
import numpy as np

# Change working directory
os.chdir('/data/results/desiree/input/')

# Load diabetes data
df_diabetes = pd.read_csv('diabetic_data.csv')
pd.set_option('display.max_columns', None)
df_diabetes.head()

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital payer_code         medical_specialty  num_lab_procedures  \
0                 1          ?  Pediatrics-Endocrinology                  41   
1                 3          ?                         ?                  59   
2                 2          ?                         ?                  11   
3                 2          ?                         ?                  44   
4                 1          ?                         ?                  51   

   num_procedures  num_medications  number_outpatient  number_emergency  \
0               0                1                  0                 0   
1               0               18                  0                 0   
2               5               13                  2                 0   
3               1               16                  0                 0   
4               0                8                  0                 0   

   number_inpatient  diag_1  diag_2 diag_3  number_diagnoses max_glu_serum  \
0                 0  250.83       ?      ?                 1           NaN   
1                 0     276  250.01    255                 9           NaN   
2                 1     648     250    V27                 6           NaN   
3                 0       8  250.43    403                 7           NaN   
4                 0     197     157    250                 5           NaN   

  A1Cresult metformin repaglinide nateglinide chlorpropamide glimepiride  \
0       NaN        No          No          No             No          No   
1       NaN        No          No          No             No          No   
2       NaN        No          No          No             No          No   
3       NaN        No          No          No             No          No   
4       NaN        No          No          No             No          No   

  acetohexamide glipizide glyburide tolbutamide pioglitazone rosiglitazone  \
0            No        No        No          No           No            No   
1            No        No        No          No           No            No   
2            No    Steady        No          No           No            No   
3            No        No        No          No           No            No   
4            No    Steady        No          No           No            No   

  acarbose miglitol troglitazone tolazamide examide citoglipton insulin  \
0       No       No           No         No      No          No      No   
1       No       No           No         No      No          No      Up   
2       No       No           No         No      No          No      No   
3       No       No           No         No      No          No      Up   
4       No       No           No         No      No          No  Steady   

  glyburide-metformin glipizide-metformin glimepiride-pioglitazone  \
0                  No                  No                       No   
1                  No                  No                       No   
2                  No                  No                       No   
3                  No                  No                       No   
4        

### Review Dataset Shape

In [31]:
# Change working directory
os.chdir('/data/results/desiree/output/')

# Dataset Rows & Columns count
# Checking number of rows and columns of the dataset using shape
print("Number of rows are: ",df_diabetes.shape[0])
print("Number of columns are: ",df_diabetes.shape[1])

Number of rows are:  101766
Number of columns are:  50


### Review Dataset Information

In [32]:
# Dataset Info
# Checking information about the dataset using info
df_diabetes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

In [33]:
# Dataset Duplicate Value Count
dup = df_diabetes.duplicated().sum()
print(f'number of duplicated rows are {dup}')

number of duplicated rows are 0


In [34]:
# Missing Values/Null Values Count
df_diabetes.isnull().sum()

encounter_id                    0
patient_nbr                     0
race                            0
gender                          0
age                             0
weight                          0
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                      0
medical_specialty               0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                          0
diag_2                          0
diag_3                          0
number_diagnoses                0
max_glu_serum               96420
A1Cresult                   84748
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

Summary of high level overview of data:
- Patient encounter data including demographic, diagnoses, medications, and readmission information
- A patient can have more than one encounter (will need to decide on how to handle for modeling)
- Dimensions: 101766 rows, 50 columns
- Columns listed as objects requiring further review to confirm consistency in reporting
    - race
    - gender
    - age
    - weight
    - payer_code
    - medical_specialty
    - diag_1, diag_2, diag_3
    - max_glu_serum (5346 non-null)
    - A1Cresult (17018 non-null)
    - medications (metformin ... metformin-pioglitazone)
    - change
    - diabetesMed
    - readmitted
- Number of duplicated rows: 0
- Columns with NULL values: max_glu_serum (96420) and A1Cresult (84748)

In [35]:
# Print column names
print(df_diabetes.columns)

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')


In [7]:
# Dataset Describe  (all columns included)
df_diabetes.describe(include='all').round(2)

encounter_id   patient_nbr       race  gender      age  weight  \
count   1.017660e+05  1.017660e+05     101766  101766   101766  101766   
unique           NaN           NaN          6       3       10      10   
top              NaN           NaN  Caucasian  Female  [70-80)       ?   
freq             NaN           NaN      76099   54708    26068   98569   
mean    1.652016e+08  5.433040e+07        NaN     NaN      NaN     NaN   
std     1.026403e+08  3.869636e+07        NaN     NaN      NaN     NaN   
min     1.252200e+04  1.350000e+02        NaN     NaN      NaN     NaN   
25%     8.496119e+07  2.341322e+07        NaN     NaN      NaN     NaN   
50%     1.523890e+08  4.550514e+07        NaN     NaN      NaN     NaN   
75%     2.302709e+08  8.754595e+07        NaN     NaN      NaN     NaN   
max     4.438672e+08  1.895026e+08        NaN     NaN      NaN     NaN   

        admission_type_id  discharge_disposition_id  admission_source_id  \
count           101766.00                 101766.00            101766.00   
unique                NaN                       NaN                  NaN   
top                   NaN                       NaN                  NaN   
freq                  NaN                       NaN                  NaN   
mean                 2.02                      3.72                 5.75   
std                  1.45                      5.28                 4.06   
min                  1.00                      1.00                 1.00   
25%                  1.00                      1.00                 1.00   
50%                  1.00                      1.00                 7.00   
75%                  3.00                      4.00                 7.00   
max                  8.00                     28.00                25.00   

        time_in_hospital payer_code medical_specialty  num_lab_procedures  \
count          101766.00     101766            101766           101766.00   
unique               NaN         18                73                 NaN   
top                  NaN          ?                 ?                 NaN   
freq                 NaN      40256             49949                 NaN   
mean                4.40        NaN               NaN               43.10   
std                 2.99        NaN               NaN               19.67   
min                 1.00        NaN               NaN                1.00   
25%                 2.00        NaN               NaN               31.00   
50%                 4.00        NaN               NaN               44.00   
75%                 6.00        NaN               NaN               57.00   
max                14.00        NaN               NaN              132.00   

        num_procedures  num_medications  number_outpatient  number_emergency  \
count        101766.00        101766.00          101766.00         101766.00   
unique             NaN              NaN                NaN               NaN   
top                NaN              NaN                NaN               NaN   
freq               NaN              NaN                NaN               NaN   
mean              1.34            16.02               0.37              0.20   
std               1.71             8.13               1.27              0.93   
min               0.00             1.00               0.00              0.00   
25%               0.00            10.00               0.00              0.00   
50%               1.00            15.00               0.00              0.00   
75%               2.00            20.00               0.00              0.00   
max               6.00            81.00              42.00             76.00   

        number_inpatient  diag_1  diag_2  diag_3  number_diagnoses  \
count          101766.00  101766  101766  101766         101766.00   
unique               NaN     717     749     790               NaN   
top                  NaN     428     276     250               NaN   
freq                 NaN    6862    6752

In [37]:
# Create a table containing mean, median, mode, and standard deviation for numerical columns
numerical_columns = df_diabetes.select_dtypes(include=np.number).columns
# Drop identifier columns
numerical_columns = numerical_columns.drop(['encounter_id', 'patient_nbr'])
statistics_table = pd.DataFrame(columns=['Column', 'Mean', 'Median', 'Mode', 'Standard Deviation'])

for column in numerical_columns:
    mean = df_diabetes[column].mean()
    median = df_diabetes[column].median()
    mode = df_diabetes[column].mode().values[0]
    std_dev = df_diabetes[column].std()
    new_row = pd.DataFrame({'Column': [column], 'Mean': [mean], 'Median': [median], 'Mode': [mode], 'Standard Deviation': [std_dev]})
    statistics_table = pd.concat([statistics_table, new_row], ignore_index=True)

statistics_table.to_csv('stats_table.csv', index=False)


**Note:**
- weight, payer_code, medical_specialty have '?' as top input

In [8]:
# Check Unique Values for each variable.
for i in df_diabetes.columns.tolist():
    print("No. of unique values in",i,"is",df_diabetes[i].nunique())

No. of unique values in encounter_id is 101766
No. of unique values in patient_nbr is 71518
No. of unique values in race is 6
No. of unique values in gender is 3
No. of unique values in age is 10
No. of unique values in weight is 10
No. of unique values in admission_type_id is 8
No. of unique values in discharge_disposition_id is 26
No. of unique values in admission_source_id is 17
No. of unique values in time_in_hospital is 14
No. of unique values in payer_code is 18
No. of unique values in medical_specialty is 73
No. of unique values in num_lab_procedures is 118
No. of unique values in num_procedures is 7
No. of unique values in num_medications is 75
No. of unique values in number_outpatient is 39
No. of unique values in number_emergency is 33
No. of unique values in number_inpatient is 21
No. of unique values in diag_1 is 717
No. of unique values in diag_2 is 749
No. of unique values in diag_3 is 790
No. of unique values in number_diagnoses is 16
No. of unique values in max_glu_seru

- 101766 encounters
- 71518 patients
- '?' Race, Weight, Payer_code, Medical_specialty (will need to clean up), Diag_1, Diag_2, Diag_3
- 'Unknown/invalid' Gender 
- 'No' only: Examide, Citoglipton

In [38]:
# Convert '?' to null values
df_diabetes.replace('?', np.nan, inplace=True)

# Create a table containing the number of missing values for each column in one column and then another column as percentage of total rows
missing_values_table = pd.DataFrame(columns=['Column', 'Missing Values', 'Percentage of Total Rows'])

for column in df_diabetes.columns:
    missing_values = df_diabetes[column].isnull().sum()
    percentage = (missing_values / len(df_diabetes)) * 100
    new_row = pd.DataFrame({'Column': [column], 'Missing Values': [missing_values], 'Percentage of Total Rows': [percentage]})
    missing_values_table = pd.concat([missing_values_table, new_row], ignore_index=True)

missing_values_table.to_csv('missing_values_table.csv', index=False)


In [9]:
# Drop columns with no variance in dataset
# These won't contribute to model prediction
df_diabetes.drop(columns=['examide', 'citoglipton'], inplace=True)
df_diabetes.shape

(101766, 48)

In [10]:
# Clean up Medical_specialty
# Updating Adult categories
df_diabetes['medical_specialty'] = df_diabetes['medical_specialty'].replace(['Gynecology', 'Obstetrics', 'ObstetricsandGynecology'], 'OBGYN')
df_diabetes['medical_specialty'] = df_diabetes['medical_specialty'].replace('Obsterics&Gynecology-GynecologicOnco', 'OBGYN/Oncology')
df_diabetes['medical_specialty'] = df_diabetes['medical_specialty'].replace('Radiologist', 'Radiology')
df_diabetes['medical_specialty'] = df_diabetes['medical_specialty'].replace('Surgery-Cardiovascular/Thoracic', 'Surgery-Cardiovascular')
df_diabetes['medical_specialty'] = df_diabetes['medical_specialty'].replace('Surgery-PlasticwithinHeadandNeck', 'Surgery-Plastic')

# Updating Pediatric categories
df_diabetes['medical_specialty'] = df_diabetes['medical_specialty'].replace('Anesthesiology-Pediatric', 'Pediatrics-Anesthesiology')
df_diabetes['medical_specialty'] = df_diabetes['medical_specialty'].replace('Cardiology-Pediatric', 'Pediatrics-Cardiology')
df_diabetes['medical_specialty'] = df_diabetes['medical_specialty'].replace('Surgery-Pediatric', 'Pediatrics-Surgery')

medical_specialty_counts = df_diabetes['medical_specialty'].value_counts()
medical_specialty_counts.to_csv('medical_specialty_counts_updated.csv')
df_diabetes.to_csv('df_diabetes_updated.csv', index=False)

Making modifications to the medical_specialty outputs for consistency and grouping of very similar categories (e.g. Radiology vs Radiologist). Could potentially lump all of Pediatrics together and Surgery together

Steps to clean up data
- need to recategorize ICD codes
- How to consolidate to represent 1 patient per row
    - Take mode: Patient_nbr, Race, Gender, Weight, Payer_code, medical_specialty, diag_1, diag_2, diag_3, max_glu_serum, A1Cresult, medications, change, diabetesMed, readmitted
    - Take most recent encounter: Age; admission_type, discharge_disposition_id, admission_source
    - Take mean / median: time_in_hospital, num_lab_procedures, num_procedures, num_medications, number_outpatient, number_emergency, number_impatient, number_diagnoses

Attempt 1 for ICD9 re-categorization
- Following the methods outlined in a paper written using this dataset, which includes 9 groups with those <3.5% of encounters lumped as 'Other'.

In [11]:
# For diag_1, diag_2, and diag_3, if the value is null, replace with '0'
for column in ['diag_1', 'diag_2', 'diag_3']:
    df_diabetes[column] = df_diabetes[column].apply(lambda x: '0' if pd.isnull(x) else x)

# Grouping similiar ICD9 diagnosis categories together
# Function to replace ICD9 diagnosis categories with group names
def replace_icd9_categories(x):
    try:
        x_int = int(x)
        if 390 <= x_int <= 459 or x == '785':
            return 'Circulatory'
        elif 460 <= x_int <= 519 or x == '786':
            return 'Respiratory'
        elif 520 <= x_int <= 579 or x == '787':
            return 'Digestive'
        elif 800 <= x_int <= 999:
            return 'Injury'
        elif 710 <= x_int <= 739:
            return 'Musculoskeletal'
        elif 140 <= x_int <= 239:
            return 'Neoplasms'
        elif x == '250':
            return 'Diabetes'
        elif 580 <= x_int <= 629 or x == '788':
            return 'Genitourinary'
        elif x_int == 0:
            return 'Unknown'
        else:
            return 'Other'
    except ValueError:
        if x.startswith('250'):
            return 'Diabetes'
        else:
            return 'Other'

# Apply the function to the diag_1 column
# For diag_1, diag_2, and diag_3, if the value is null, replace with '0'
for column in ['diag_1', 'diag_2', 'diag_3']:
    df_diabetes[f"{column}_group"] = df_diabetes[column].apply(replace_icd9_categories)

df_diabetes.to_csv('df_diabetes_icd_groups.csv', index=False)



In [20]:
# Consolidate to represent 1 patient per row
# Take mode: Patient_nbr, Race, Gender, Weight, Payer_code, medical_specialty, diag_1, diag_2, diag_3, max_glu_serum, A1Cresult, medications, change, diabetesMed, readmitted
# Take most recent encounter: Age; admission_type, discharge_disposition_id, admission_source
# Take mean: time_in_hospital, num_lab_procedures, num_procedures, num_medications, number_outpatient, number_emergency, number_impatient, number_diagnoses

# For each Patient_nbr, come up with a consensus
# Sort the DataFrame by patient_nbr and encounter_id (assuming encounter_id is a proxy for the encounter date)
df_diabetes_sorted = df_diabetes.sort_values(by=['patient_nbr', 'encounter_id'], ascending=[True, False])

# Drop duplicates to keep only the most recent encounter for each patient
df_most_recent_encounter = df_diabetes_sorted.drop_duplicates(subset=['patient_nbr'], keep='first')

# Extract the most recent age for each patient
df_most_recent_age = df_most_recent_encounter[['patient_nbr', 'age', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id']]

# Define a function to get the mode of a series
def get_mode(series):
    return series.mode().iloc[0] if not series.mode().empty else None

# Group by patient_nbr and apply the get_mode function
df_patient_info = df_diabetes.groupby('patient_nbr').agg({
    'race': lambda x: get_mode(x),
    'gender': lambda x: get_mode(x),
    'weight': lambda x: get_mode(x),
    'payer_code': lambda x: get_mode(x),
    'medical_specialty': lambda x: get_mode(x),
    'diag_1_group': lambda x: get_mode(x),
    'diag_2_group': lambda x: get_mode(x),
    'diag_3_group': lambda x: get_mode(x),
    'max_glu_serum': lambda x: get_mode(x),
    'A1Cresult': lambda x: get_mode(x),
    'change': lambda x: get_mode(x),
    'diabetesMed': lambda x: get_mode(x),
    'readmitted': lambda x: get_mode(x),
    'metformin': lambda x: get_mode(x),
    'repaglinide': lambda x: get_mode(x),
    'nateglinide': lambda x: get_mode(x),
    'chlorpropamide': lambda x: get_mode(x),
    'glimepiride': lambda x: get_mode(x),
    'acetohexamide': lambda x: get_mode(x),
    'glipizide': lambda x: get_mode(x),
    'glyburide': lambda x: get_mode(x),
    'tolbutamide': lambda x: get_mode(x),
    'pioglitazone': lambda x: get_mode(x),
    'rosiglitazone': lambda x: get_mode(x),
    'acarbose': lambda x: get_mode(x),
    'miglitol': lambda x: get_mode(x),
    'troglitazone': lambda x: get_mode(x),
    'tolazamide': lambda x: get_mode(x),
    'insulin': lambda x: get_mode(x),
    'glyburide-metformin': lambda x: get_mode(x),
    'glipizide-metformin': lambda x: get_mode(x),
    'glimepiride-pioglitazone': lambda x: get_mode(x),
    'metformin-rosiglitazone': lambda x: get_mode(x),
    'metformin-pioglitazone': lambda x: get_mode(x),
    'time_in_hospital': 'mean',
    'num_lab_procedures': 'mean',
    'num_procedures': 'mean',
    'num_medications': 'mean',
    'number_outpatient': 'mean',
    'number_emergency': 'mean',
    'number_inpatient': 'mean',
    'number_diagnoses': 'mean'
}).reset_index()

# Merge the most recent age information with the aggregated data
df_final = df_patient_info.merge(df_most_recent_age, on='patient_nbr', how='left')
# takes almost 8 min to run

In [24]:
# check for missing values
# print(df_final.isnull().sum())
# only 2 columns have >90% missing values (max_glu_serum, A1Cresult)
# drop columns with missing values
df_final.dropna(inplace=True)
print(df_final.isnull().sum())
# Save the result to a CSV file
df_final.to_csv('df_patient_info_one_row.csv', index=False)

patient_nbr                 0
race                        0
gender                      0
weight                      0
payer_code                  0
medical_specialty           0
diag_1_group                0
diag_2_group                0
diag_3_group                0
max_glu_serum               0
A1Cresult                   0
change                      0
diabetesMed                 0
readmitted                  0
metformin                   0
repaglinide                 0
nateglinide                 0
chlorpropamide              0
glimepiride                 0
acetohexamide               0
glipizide                   0
glyburide                   0
tolbutamide                 0
pioglitazone                0
rosiglitazone               0
acarbose                    0
miglitol                    0
troglitazone                0
tolazamide                  0
insulin                     0
glyburide-metformin         0
glipizide-metformin         0
glimepiride-pioglitazone    0
metformin-

In [ ]:
# Create a table containing mode for categorical columns
categorical_columns = df_diabetes.select_dtypes(include='object').columns
mode_table = pd.DataFrame(columns=['Column', 'Mode'])

for column in categorical_columns:
    mode = df_diabetes[column].mode().values[0]
    new_row = pd.DataFrame({'Column': [column], 'Mode': [mode]})
    mode_table = pd.concat([mode_table, new_row], ignore_index=True)

print(mode_table)